### PR TITLE
[DirectX] Clean up extra vectors when lowering to buffer store

### PR DESCRIPTION
**Pull request by bogner**
DXILOpLowering runs after scalarization but `@llvm.dx.typedbuffer.store` takes a vector, so the argument is usually an artifact. Avoid creating a vector just to extract elements from it immediately.

### CODE REVIEW

File: llvm/lib/Target/DirectX/DXILOpLowering.cpp [hunk 1]
```diff
@@ -531,23 +531,46 @@ class OpLowerer {
...
-      Value *Data2 =
-          IRB.CreateExtractElement(Data, ConstantInt::get(Int32Ty, 2));
-      Value *Data3 =
-          IRB.CreateExtractElement(Data, ConstantInt::get(Int32Ty, 3));
-
-      std::array<Value *, 8> Args{Handle, Index0, Index1, Data0,
-                                  Data1,  Data2,  Data3,  Mask};
+
+      // Since we're post-scalarizer, we likely have a vector that's constructed
+      // solely for the argument of the store.
```
Comment (damyanp) [inline llvm/lib/Target/DirectX/DXILOpLowering.cpp:536]: FWIW, I feel that there's a "...so we can <do whatever this loop does>" missing from this comment, but that might just be my lack of familiarity with this sort of code speaking.

File: llvm/lib/Target/DirectX/DXILOpLowering.cpp [hunk 2]
```diff
@@ -531,23 +531,46 @@ class OpLowerer {
...
+      std::array<Value *, 4> DataElements{nullptr, nullptr, nullptr, nullptr};
+      auto *IEI = dyn_cast<InsertElementInst>(Data);
+      while (IEI) {
+        auto *IndexOp = dyn_cast<ConstantInt>(IEI->getOperand(2));
+        if (!IndexOp)
+          break;
+        size_t IndexVal = IndexOp->getZExtValue();
+        assert(IndexVal < 4 && "Too many elements for buffer store");
```
Comment (damyanp) [inline llvm/lib/Target/DirectX/DXILOpLowering.cpp:544]: If we start support > 4 element vectors, are asserts like this going to bite us (ie hard to debug errors that don't show up in production builds?)

**Review by damyanp — COMMENTED** [review]
Looks Good To My Untrained Eye

File: llvm/lib/Target/DirectX/DXILOpLowering.cpp (hunk 1 above)
Comment (bogner) [reply llvm/lib/Target/DirectX/DXILOpLowering.cpp]: Added "If so, just use the scalar values from before they're inserted into the temporary." to the comment.

File: llvm/lib/Target/DirectX/DXILOpLowering.cpp (hunk 2 above)
Comment (bogner) [reply llvm/lib/Target/DirectX/DXILOpLowering.cpp:545]: I don't think so, this is fairly likely to crash in a non-asserts build if it comes up anyways, and this is a fairly obvious place we'd need to update if we were to make such a change. Note that we'd have to expand into multiple calls to `dx.op.BufferStore` here if we were to support larger vectors in HLSL, so this isn't just some implementation detail we'd be likely to forget about.

**Review by llvm-beanz — APPROVED** [review]

**Pull request merged**